pyClient/test_commands/scenario.py

```diff
@@ -153,6 +155,45 @@ def charlie_double_withdraw(
     note1_value = joinsplit.to_zeth_units(EtherValue(CHARLIE_WITHDRAW_CHANGE_ETH))
     v_out = EtherValue(CHARLIE_WITHDRAW_ETH)
 
+    # ### ATTACK BLOCK
+    # Add malicious nullifiers: we reuse old nullifiers to double spend by
+    # adding $r$ to them so that they have the same value as before in Z_r,
+    # and so the zksnark verification passes, but have different values in
+    # {0;1}^256 so that they appear different to the contract.
+    # See: https://github.com/clearmatics/zeth/issues/38
+
+    def compute_h_sig_attach_nf(
+            nf0: bytes,
+            nf1: bytes,
+            sign_pk: signing.SigningPublicKey) -> bytes:
+        r = 21888242871839275222246405745257275088548364400416034343698204186575808495617  # noqa
+
+        # We disassemble the nfs to get the formatting of the primary inputs
+        input_nullifier0 = nf0.hex()
+        input_nullifier1 = nf1.hex()
+        nf0_rev = "{0:0256b}".format(int(input_nullifier0, 16))[::-1]
+        primary_input1_bits = nf0_rev[3:]
+        primary_input2_bits = nf0_rev[:3]
+        nf1_rev = "{0:0256b}".format(int(input_nullifier1, 16))[::-1]
+        primary_input3_bits = nf1_rev[3:]
+        primary_input4_bits = nf1_rev[:3]
+
+        # We perform the attack
+        attack_primary_input2 = int(primary_input2_bits, 2) + r
+        attack_primary_input4 = int(primary_input4_bits, 2) + r
```

Comment:
Why not pass the "mauled" nullifiers as arguments to the function directly?

In this function, you redefine the prime `r`, and do
```python
# We perform the attack
attack_primary_input2 = int(primary_input2_bits, 2) + r
attack_primary_input4 = int(primary_input4_bits, 2) + r
```

Which is done also a few lines later (outside of you callback)
```python
proof_json["inputs"][2] = hex(int(proof_json["inputs"][2], 16) + r)
proof_json["inputs"][4] = hex(int(proof_json["inputs"][4], 16) + r)
# ### ATTACK BLOCK
```

Reply:
I've checked to use ZETH_PRIME and cached the public input values in the first instance.

The nullifier and h_sig are computed internally in the joinsplit module, hence the need for a callback instead of passing the computed values somewhere.  This avoids having extra fucntions in joinsplit which are purely for testing.